pkg/imagesize/imagesize.go

```diff
@@ -0,0 +1,164 @@
+package imagesize
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/reference/docker"
+	manifesttypes "github.com/docker/cli/cli/manifest/types"
+	"github.com/docker/cli/cli/registry/client"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/reference"
+	registrytypes "github.com/docker/docker/api/types/registry"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	typedv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	cri "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+)
+
+const (
+	ctxTimeout = 2 * time.Minute
+)
+
+type warner struct {
+	recorder     record.EventRecorder
+	broadcaster  record.EventBroadcaster
+	clientSet    kubernetes.Interface
+	maxImageSize *resource.Quantity
+}
+
+var Warner *warner
+
+// Initialize initializes the singleton `Warner` object
+func Initialize(maxImageSize *resource.Quantity) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		klog.Fatalf("unable to get cluster config: %s", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		klog.Fatalf("unable to get cluster client: %s", err)
+	}
+
+	scheme := runtime.NewScheme()
+	_ = v1.AddToScheme(scheme)
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartStructuredLogging(1)
+	eventBroadcaster.StartRecordingToSink(&typedv1.EventSinkImpl{Interface: clientset.CoreV1().Events("")})
+	eventRecorder := eventBroadcaster.NewRecorder(scheme, v1.EventSource{
+		Component: "warm-metal",
+	})
+
+	Warner = &warner{
+		recorder:     eventRecorder,
+		broadcaster:  eventBroadcaster,
+		clientSet:    clientset,
+		maxImageSize: maxImageSize,
+	}
+}
+
+// Cleanup performs cleanup activity (like shutting down the event broadcaster)
+func (w *warner) Cleanup() {
+	w.broadcaster.Shutdown()
+}
+
+// Warn creates the warning event object for the pod
+func (w *warner) Warn(current *resource.Quantity, pod string, namespace string) error {
+	if current.Cmp(*w.maxImageSize) <= 0 {
+		return nil
+	}
```

Comment:
Is this where we check if `maxImageSize` flag is set, and if not then returning without errors?

Reply:
If yes, then maybe we should add this check early before calling GetImageSize() function. That would reduce the additional work on the driver if the flag is not enabled.